File: solutions/day07/solution.py
import re


class Switch(object):
    operator_to_fn = {'AND': lambda a, b: (a & b) & 0xFFFF,
                      'OR': lambda a, b: (a | b) & 0xFFFF,
                      'NOT': lambda a: ~a & 0xFFFF,
                      'LSHIFT': lambda a, b: (a << b) & 0xFFFF,
                      'RSHIFT': lambda a, b: (a >> b) & 0xFFFF,
                      'IDENTITY': lambda a: a & 0xFFFF}

    def __init__(self, name, operator, inputs, circuit):
        self.name = name
        self.operator = operator
        self.value = None

        def make_input_fn(i):
            if re.match('[0-9]+', i):
                return lambda: int(i)
            else:
                return lambda: circuit[i].value

        self.inputs = map(make_input_fn, inputs)

    def reset(self):
        self.value = None

    def set_output(self):
        input_values = [i() for i in self.inputs]
        if all([i is not None for i in input_values]):
            self.value = self.operator_to_fn[self.operator](*input_values)


class Circuit(dict):
    def __init__(self, rules_file='input.txt'):
        self.load(rules_file)

    def load(self, rules_file='input.txt'):
        with open(rules_file, 'r') as fh:
            for line in fh:
                instruction, target = line.strip().split(' -> ')
                parts = instruction.split(' ')
                if len(parts) == 1:
                    operator = 'IDENTITY'
                    inputs = parts
                elif len(parts) == 2:
                    operator = parts[0]
                    inputs = [parts[1]]
                elif len(parts) == 3:
                    operator = parts[1]
                    inputs = [parts[0], parts[2]]

                switch = Switch(target, operator, inputs, self)
                self[target] = switch

    def energize(self):
        while not all(s.value is not None for s in self.values()):
            for switch in self.values():
                if switch.value is None:
                    switch.set_output()

    def reset(self):
        for switch in self.values():
            switch.reset()


circuit = Circuit('input.txt')
circuit.energize()
print 'part 1:', circuit['a'].value

tmp = circuit['a'].value
circuit.reset()
circuit['b'].value = tmp
circuit.energize()
print 'part 2:', circuit['a'].value

File: solutions/day07/solution.clj
(require '[clojure.string :as str])

(def cmd->fn {"OR" #(bit-or %1 %2)
              "AND" #(bit-and %1 %2)
              "RSHIFT" #(bit-shift-right %1 %2)
              "LSHIFT" #(bit-shift-left %1 %2)
              "NOT" #(bit-not %)
              "SET" #(identity %)})

(defn make-wire
  [x]
  (fn [circuit]
    (if (re-find #"^[0-9]+$" x)
      (read-string x)
      (circuit x))))

(defn instruction->switch 
  [instruction]
  (let [[input target] (str/split instruction #" -> ")
        parts (str/split input #" ")
        commands (set (keys cmd->fn))
        cmd (or (first (filter commands parts)) "SET")
        operands (map make-wire (remove commands parts))]
    {:inputs (fn [circuit] (map #(% circuit) operands))
     :fn (cmd->fn cmd)
     :target target}))

(def switches (->> (slurp "input.txt")
                   (str/split-lines)
                   (map instruction->switch)))

(defn switches->circuit 
  ([switches] (switches->circuit switches {}))
  ([switches init-circuit]
    (loop [[switch & more] switches
           circuit init-circuit]
      (if switch
        (let [inputs ((:inputs switch) circuit)]
          (if (not-any? nil? inputs)
            (recur more (assoc circuit (:target switch) (apply (:fn switch) inputs)))
            ; inputs for this switch aren't set, so put it at the end of the list 
            (recur (conj (vec more) switch) circuit)))
        circuit))))

(defn part1
  [switches]
  ((switches->circuit switches) "a"))

(defn part2
  [switches init-circuit]
  (let [set-switches (set (keys init-circuit))
        switches (filter #(not (set-switches (:target %))) switches)]
    ((switches->circuit switches init-circuit) "a")))

(println "part 1:" (part1 switches))
(println "part 2:" (part2 switches {"b" (part1 switches)}))
